Benchmarks/Benchmarks/Calendar/Calendar.swift

```diff
@@ -40,7 +40,26 @@ let benchmarks = {
             }
         }
     }
-
+    Benchmark("nextThousandThanksgivingsSequence") { benchmark in
+        var count = 1000
+        for _ in cal.dates(byMatching: thanksgivingComponents, startingAt: thanksgivingStart, matchingPolicy: .nextTime) {
+            count -= 1
+            if count == 0 {
+                break
+            }
+        }
+    }
+    Benchmark("nextThousandThanksgivingsUsingRecurrenceRule") { benchmark in
+        var rule = Calendar.RecurrenceRule(calendar: cal, frequency: .yearly, end: .afterOccurrences(1000))
+        rule.months = [11]
+        rule.weekdays = [.nth(4, .thursday)]
+        rule.matchingPolicy = .nextTime
+        var count = 0
+        for _ in rule.recurrences(of: thanksgivingStart) {
+            count += 1
+        }
+        assert(count == 1000)
```

Comment:
I fixed an obvious performance pitfall (we were enumerating over a wider range than needed), and now we are within <4x runtime of the benchmark that uses enumeration by date components. Since we also own the implementation for matching date components, the next step would be to bypass the public API and use some of the internals to get the runtime even closer.

Reply:
The original benchmark (`nextThousandThanksgivings`) does not actually compute Thanksgivings as the name implies. It finds the Thursday in the fourth week of November, which is not necessarily the fourth Thursday in November. Is it permissible to update the benchmark given that the runtime will be slightly different?

Reply:
> Since we also own the implementation for matching date components, the next step would be to bypass the public API and use some of the internals to get the runtime even closer.

I got started on this, but the potential performance improvement does not seem to justify the amount of work it takes to get there. I'd like to merge the implementation as it is now, and revisit performance with a later PR. The benchmarks taken on my machine are as follows:

```
nextThousandThanksgivings
╒═══════════════════════════╤═════════╤═════════╤═════════╤═════════╤═════════╤═════════╤═════════╤═════════╕
│ Metric                    │      p0 │     p25 │     p50 │     p75 │     p90 │     p99 │    p100 │ Samples │
╞═══════════════════════════╪═════════╪═════════╪═════════╪═════════╪═════════╪═════════╪═════════╪═════════╡
│ Malloc (total) *          │     426 │     426 │     426 │     426 │     426 │     426 │     426 │       6 │
├───────────────────────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┤
│ Throughput (# / s) (K)    │       2 │       2 │       2 │       2 │       2 │       2 │       2 │       6 │
├───────────────────────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┤
│ Time (total CPU) (μs) *   │     580 │     581 │     581 │     583 │     583 │     583 │     583 │       6 │
├───────────────────────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┤
│ Time (wall clock) (μs) *  │     582 │     584 │     584 │     586 │     586 │     586 │     586 │       6 │
╘═══════════════════════════╧═════════╧═════════╧═════════╧═════════╧═════════╧═════════╧═════════╧═════════╛

nextThousandThanksgivingsSequence
╒═══════════════════════════╤═════════╤═════════╤═════════╤═════════╤═════════╤═════════╤═════════╤═════════╕
│ Metric                    │      p0 │     p25 │     p50 │     p75 │     p90 │     p99 │    p100 │ Samples │
╞═══════════════════════════╪═════════╪═════════╪═════════╪═════════╪═════════╪═════════╪═════════╪═════════╡
│ Malloc (total) *          │     426 │     426 │     426 │     426 │     426 │     426 │     426 │       6 │
├───────────────────────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┤
│ Throughput (# / s) (K)    │       2 │       2 │       2 │       2 │       2 │       2 │       2 │       6 │
├───────────────────────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┤
│ Time (total CPU) (μs) *   │     580 │     581 │     581 │     581 │     581 │     581 │     581 │       6 │
├───────────────────────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┤
│ Time (wall clock) (μs) *  │     582 │     584 │     584 │     585 │     586 │     586 │     586 │       6 │
╘═══════════════════════════╧═════════╧═════════╧═════════╧═════════╧═════════╧═════════╧═════════╧═════════╛

nextThousandThanksgivingsUsingRecurrenceRule
╒═══════════════════════════╤═════════╤═════════╤═════════╤═════════╤═════════╤═════════╤═════════╤═════════╕
│ Metric                    │      p0 │     p25 │     p50 │     p75 │     p90 │     p99 │    p100 │ Samples │
╞═══════════════════════════╪═════════╪═════════╪═════════╪═════════╪═════════╪═════════╪═════════╪═════════╡
│ Malloc (total) *          │     861 │     861 │     861 │     861 │     861 │     861 │     861 │       2 │
├───────────────────────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┤
│ Time (total CPU) (μs) *   │    2121 │    2121 │    2121 │    2129 │    2129 │    2129 │    2129 │       2 │
├───────────────────────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┤
│ Time (wall clock) (μs) *  │    2135 │    2136 │    2136 │    2140 │    2140 │    2140 │    2140 │       2 │
╘═══════════════════════════╧═════════╧═════════╧═════════╧═════════╧═════════╧═════════╧═════════╧═════════╛
```

Reply:
Ok, that seems reasonable for now. Let's file a follow-up bug just so we don't lose track of the task.

Reply:
Tracking that in #555 (rdar://126761035)